R/read-gs.R

```diff
@@ -18,14 +19,225 @@
+                       other_file_paths = list(), 
+                       source = "auto"){
+  # first read in all of the files
+  
+  # beginning with the grades csv
+  
+  
+    grades <- readr::read_csv(grades_path, trim_ws = FALSE)
+    
+  
+  # now read in other supplied data 
+ for (data_type in names(other_file_paths)){
+   warning(paste0(data_type, 
+                  " is not currently supported as an extra file source."))
+ }
+  
+  # now determine how to read in data
+  
+  auto_determine <- determine_grade_source(grades)
+  
+  if (source == "auto"){
+    
+    
+    if (auto_determine == "Gradescope"){
+      
+      
+      return(read_gradescope_grades(grades))
+    } else if (auto_determine == "Canvas"){
+      # the read_canvas_grades does all of the lifting
+      return(read_canvas_grades(grades))
+    } else {
+      stop("Data source is unrecognized.")
+    }
+    
+  } else if (source == "Gradescope"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    
+    
+    
+    return(read_gradescope_grades(grades))
+      
+  } else if (source == "Canvas"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    # the read_canvas_grades does all of the lifting
+    return(read_canvas_grades(grades))
+    
+  } else {
+    stop(paste0(
+      "Specified source type ", source,
+      " is not currently supported."
+    ))
+  }
+  
+  
+ 
+}
+
+read_gradescope_grades <- function(df){
+  # almost identity function when reading in gradescope grades 
+  # to create symmetry with read_canvas_grades
+  
+  attr(df, "source") <- "Gradescope"
+  check_data_format(df)
+}
+
+
+
+#' @importFrom stringr str_extract str_match str_replace
+#' @importFrom dplyr filter select slice rename_with rename bind_cols mutate_at
+#' @importFrom purrr discard
+#' @importFrom lubridate make_difftime
+#' 
+read_canvas_grades <- function(grades){
+  # Convert Canvas grade dataframe into standard format
+  #
+  # Takes in a dataframe uploaded from Canvas. 
+  # Returns a dataframe in our standardized format (ie like Gradescope).
+  #
+  #
+  
+  # first determine which columns are assignments
+  
+  # canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+  # coerce into numerical and convert excused assignments into set value
+  grades <- grades |>
+    dplyr::filter(!is.na(Student)) |>
+    dplyr::mutate_at(assignments, function(x) {
+        dplyr::if_else(x == "EX", NA, x) |>
```

Comment:
We read in assignment scores marked as NA as zeroes (our reasoning here is that a not-turned-in assignment is the same as having a zero on a turned-in assignment)... I'm not familiar with excused assignments but I don't know if this is the right coercion